src/views/common/Animation/AnimationStateReducer.tsx

```diff
@@ -0,0 +1,231 @@
+            return setFrame(s, a)
+        case SET_REPLAY_RATE:
+            // Negative values are allowed and will just play it backwards.
+            if (a.newRate === 0) {
+                console.warn(`Attempt to set new replay rate to 0.`)
+                return s
+            }
+            refreshAnimationCycle(s)
+            return { ...s, framesPerTick: a.newRate }
+        case SKIP:
+            // Currently implemented as a jump of 3 pct, but that's just a placeholder; maybe fixed length (or fixed time) would be better?
+            // Thought: This could interact with the replay rate and skip e.g. 5 seconds of playback time under current settings...?
+            const newFrame = s.currentFrameIndex + (Math.floor(s.frameData.length * .03) * (a.backward ? -1 : 1))
+            s.currentFrameIndex = newFrame < 0
+                ? 0
+                : newFrame > (s.frameData.length - 1)
+                    ? s.frameData.length : newFrame
+            refreshAnimationCycle(s)
+            return {...s}
+        case TO_END:
+            s.currentFrameIndex = a.backward ? 0 : s.frameData.length - 1
+            refreshAnimationCycle(s)
+            return {...s}
+        default: {
+            throw Error(`Invalid action type for animation state reducer: ${type}`)
+        }
+    }
+}
+
+const refreshAnimationCycle = (s: AnimationState<any>) => {
+    if (s.pendingFrameCode === undefined || !s.isPlaying) return
+    window.cancelAnimationFrame(s.pendingFrameCode)
+    if (s.animationDispatchFn === undefined) {
+        console.warn('Animation callback unset.')
+        return
+    }
+    s.pendingFrameCode = window.requestAnimationFrame(s.animationDispatchFn)
+}
+
+
+const updateFrames = <T, >(s: AnimationState<T>, a: AnimationStateUpdateFrameDataAction<T>): AnimationState<T> => {
+    if (a.replaceExistingFrames) {
+        s.frameData = a.incomingFrames
+    } else {
+        // apparently iteratively inserting these is orders of magnitude faster than using the concat function,
+        // but the reference is pretty old, so we should consider profiling this
+        a.incomingFrames.forEach(frame => s.frameData.push(frame))
+    }
+    return {...s}
+}
+
+
+const MS_PER_TICK = 1000/60
+const doTick = <T, >(s: AnimationState<T>, a: AnimationStateTickAction): AnimationState<T> => {
+    if (s.animationDispatchFn === undefined) {
+        console.warn('Attempt to animate, but no dispatch function set. No-op.')
+        return s
+    }
+
+    if (s.lastRenderedTimestamp) {
+        const elapsedMs = a.now - s.lastRenderedTimestamp
+        if (elapsedMs < 0) {
+            console.warn(`Doing animation tick, but elapsed time is somehow negative: ${a.now} - ${s.lastRenderedTimestamp} = ${elapsedMs}`)
+            return s
+        }
+        const rawFramesToAdvance = elapsedMs/MS_PER_TICK * s.framesPerTick
+        const framesToAdvance = s.framesPerTick > 0 ? Math.floor(rawFramesToAdvance) : Math.ceil(rawFramesToAdvance)
```

Comment:
Of course, if we're going backwards, we need to CEIL rather than FLOOR or we animate too fast.